rosidl_parser/rosidl_parser/__init__.py

```diff
@@ -705,3 +709,75 @@ def parse_service_string(pkg_name, srv_name, message_string):
+        index for index, line in enumerate(lines) if line == SERVICE_REQUEST_RESPONSE_SEPARATOR]
+    if not separator_indices:
+        raise InvalidServiceSpecification(
+            "Could not find separator '%s' between action goal, result and feedback services" %
+            SERVICE_REQUEST_RESPONSE_SEPARATOR)
+    if len(separator_indices) != 3:
+        raise InvalidServiceSpecification(
+            "Could not find unique separator '%s' between action goal, result and \
+            feedback services" %
+            SERVICE_REQUEST_RESPONSE_SEPARATOR)
+
+    services = []
+    #----------------------------------------------------------------------------------------------
+    ## Send goal
+    implicit_input = ["uint8[16] uuid"]
+    request_message_string = '\n'.join(lines[:separator_indices[0]] + implicit_input)
+    request_message = parse_message_string(
+        pkg_name,
+        action_name + ACTION_GOAL_MESSAGE_SUFFIX + SERVICE_REQUEST_MESSAGE_SUFFIX,
+        request_message_string)
+
+    implicit_output = ["bool accepted", "builtin_interfaces/Time stamp"]
+    response_message_string = '\n'.join(implicit_output)
+    response_message = parse_message_string(
+        pkg_name,
+        action_name + ACTION_GOAL_MESSAGE_SUFFIX + SERVICE_RESPONSE_MESSAGE_SUFFIX,
+        response_message_string)
+
+    services.append(ServiceSpecification(pkg_name, action_name, request_message, response_message))
+    #----------------------------------------------------------------------------------------------
+
+    #----------------------------------------------------------------------------------------------
+    ## Get result
+    implicit_input = ["uint8[16] uuid"]
+    request_message_string = '\n'.join(implicit_input)
+    request_message = parse_message_string(
+        pkg_name,
+        action_name + ACTION_RESULT_MESSAGE_SUFFIX + SERVICE_REQUEST_MESSAGE_SUFFIX,
+        request_message_string)
+
+    implicit_output = ["int8 status"]
+    response_message_string = '\n'.join(
+        lines[separator_indices[0] + 1 : separator_indices[1]] + implicit_output)
+    response_message = parse_message_string(
+        pkg_name,
+        action_name + ACTION_RESULT_MESSAGE_SUFFIX + SERVICE_RESPONSE_MESSAGE_SUFFIX,
+        response_message_string)
+
+    services.append(ServiceSpecification(pkg_name, action_name, request_message, response_message))
+    #----------------------------------------------------------------------------------------------
+
+    #----------------------------------------------------------------------------------------------
+    ## Feedback message
+    implicit_input = ["uint8[16] uuid"]
+    message_string = '\n'.join(
+        lines[separator_indices[1] + 1:] + implicit_output)
+    feedback_msg = parse_message_string(
+        pkg_name, action_name + ACTION_FEEDBACK_MESSAGE_SUFFIX, message_string)
+    #----------------------------------------------------------------------------------------------
```

Comment:
@sservulo Hmm, implicit fields may clash with explicitly given ones. Action generation in ROS1 would nest message definition as a field of a completely implicit message, see [genaction.py](docs.ros.org/melodic/api/actionlib_msgs/html/genaction_8py_source.html) sources. Alternatively, we could check for the presence of duplicate entries. @sloretz @jacobperron thoughts?

Reply:
That's a good point. +1 to nesting the user defined message

Reply:
Other issues were already adressed. @sloretz, this one will probably delay a bit cmake header typesupport on #302 and #303 due to the need to update location assumptions of nested msgs.